soc/xtensa/intel_adsp/common/include/soc.h

```diff
@@ -136,12 +136,16 @@ static ALWAYS_INLINE void z_idelay(int n)
 /* memcopy used by boot loader */
 static ALWAYS_INLINE void bmemcpy(void *dest, void *src, size_t bytes)
 {
+	#ifdef __cplusplus
+	uint32_t *d = reinterpret_cast<uint32_t *>(dest);
+	uint32_t *s = reinterpret_cast<uint32_t *>(src);
```

Comment:
please, use C style casts and remove this ifdef.
```
uint32_t *d = (uint32_t *)dest;
uint32_t *s = (uint32_t *)src;
```

Reply:
No, I thought about this cast for a while and reinterpret_cast is best solution here.
In case of C++ there is no reason to use inefficient C-cast style.
As a reminder, C++ C-cast style performs different operation than in C.
It does following:
1. tries: const_cast
2. tries: static_cast (which is not desirable in this case as this code is invoked from C and literally anything can be passed as source and destination. Most of the casts would fit to this case however).
3. tries static_cast , then const_cast (change type + remove const)
4. tries reinterpret_cast (the one I used here).
5. tries reinterpret_cast , then const_cast (change type + remove const)

Reference: https://en.cppreference.com/w/cpp/language/explicit_cast

To sum up:
`uint32_t *d = (uint32_t *)dest;`
introduced overhead of:
99% of cases - 1 extra cast attempt (const cast)
1% of cases - 3 extra casts attempts

That being said, C-cast is considered a bad practice in C++.
Some explanation: https://isocpp.org/wiki/faq/coding-standards#coding-stds-cpp-vs-c
Some more explanation: https://isocpp.org/wiki/faq/coding-standards#pointer-casts

Reply:
I understand writing C code in C++ is bad and that C-style casts are a complicated mess in C++ but does any of this "overhead" happen at run time? I don't see `dynamic_cast` in this list. In other words, does this make any difference to the generated binary? I just want to make sure I understand what you mean by "overhead".

BTW:
- https://github.com/zephyrproject-rtos/zephyr/issues/45785

Reply:
Ok I will use C-cast for now. C++ style is not implemented anyways and who cares about compile time overhead ;-)

Reply:
Interestingly, Bjarne Stroustrup recommends `uint32_t *d = static_cast<uint32_t *>(dest);` instead in section 5.6 of his C++ book https://eli.thegreenplace.net/2009/11/16/void-and-casts-in-c-and-c

(I opened the book, found section 5.6 and successfully fact-checked this page)

This is what's great with C++: spoiling us with choice :-P

I wonder whether the real question that the various compilers are trying to ask us is maybe this: why does this function accepts `void *` when it actually supports only `uint32_t *`? This looks like a bug...

Reply:
Well normally in only C++ clean code we obviously should use static cast in this case as you said.
My concerns are related to interopterability with C where we clearly can pass anything as void* to the function.